src/inmanta/agent/resourcepool.py

```diff
@@ -0,0 +1,395 @@
+
+import inmanta.util
+from inmanta.const import LOG_LEVEL_TRACE
+
+LOGGER = logging.getLogger(__name__)
+
+TPoolID = TypeVar("TPoolID")
+TIntPoolID = TypeVar("TPoolID")
+
+
+class PoolMember(abc.ABC, Generic[TPoolID]):
+    """
+    Item that can live in a pool
+
+    The lifecycle if this item is:
+    - running (after construction_
+    - shutting down (after call to request_shutdown)
+    - shutdown (after call to set_shutdown)
+
+    Implementors of sub-classes are expected to
+    - override request_shutdown to start the shutdown of the underlying resource
+    - at the end of request_shutdown, call set_shutdown to indicate we are down
+
+    After request_shutdown is called, this member will be considered invalid.
+        When checked out of any pool a replacement will be produced.
+    When set_shutdown is called, the shutdown event will be sent to all pools containing this member.
+        This will cause it to be evicted
+    """
+
+    def __init__(self, my_id: TPoolID):
+        self.id = my_id
+
+        # Time based expiry
+        self._last_used: datetime.datetime = datetime.datetime.now().astimezone()
+
+        # state tracking
+        self.shutting_down = False
+        self.shut_down = False
+
+        # event propagation
+        self.termination_listeners: list[Callable[[PoolMember[TPoolID]], Coroutine[Any, Any, Any]]] = []
+
+    @property
+    def running(self) -> bool:
+        return not self.shutting_down and not self.shut_down
+
+    @property
+    def last_used(self) -> datetime.datetime:
+        return self._last_used
+
+    def touch(self) -> None:
+        """Update time last used"""
+        self._last_used = datetime.datetime.now().astimezone()
+
+    def can_be_cleaned_up(self) -> bool:
+        """
+        Return true if this member is not inhibited from cleanup
+        """
+        return True
+
+    def get_idle_time(self) -> datetime.timedelta:
+        """
+        Retrieve the idle time of this pool member
+        """
+        return datetime.datetime.now().astimezone() - self.last_used
+
+    def get_id(self) -> TPoolID:
+        """
+        Returns the ID of the pool member that will be used to lock the pool member. This ensures that no operations will
+        overlap while this pool member is being created, modified, or cleaned.
+        """
+        return self.id
+
+    async def request_shutdown(self) -> None:
+        """
+        Close the pool member.
+
+        This should eventually cause a call to closed
+        """
+        self.shutting_down = True
+
+    async def set_shutdown(self) -> None:
+        """
+        This pool member is now closed
+        """
+        self.shut_down = True
+        self.shutting_down = True
+        for listener in self.termination_listeners:
+            await listener(self)
+
+
+TPoolMember = TypeVar("TPoolMember", bound=PoolMember)
+
+
+class PoolManager(abc.ABC, Generic[TPoolID, TIntPoolID, TPoolMember]):
```

Comment:
I failed to type this in a way that binds the pool member and the ids entirely correctly